---
title: "Benchmark single cell long-read"
subtitle: "visualization"
author: "Ali Hamraoui"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: show
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---

<style>
body {
text-align: justify}
</style>

<!-- Automatically computes and prints in the output the running time for any code chunk -->
```{r, echo=FALSE}
# https://github.com/rstudio/rmarkdown/issues/1453
hooks = knitr::knit_hooks$get()
hook_foldable = function(type) {
  force(type)
  function(x, options) {
    res = hooks[[type]](x, options)
    
    if (isFALSE(options[[paste0("fold_", type)]])) return(res)
    
    paste0(
      "<details><summary>", "show", "</summary>\n\n",
      res,
      "\n\n</details>"
    )
  }
}
knitr::knit_hooks$set(
  output = hook_foldable("output"),
  plot = hook_foldable("plot")
)
```

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE, # for sessionInfo()
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15)
```


```{r locations, message=FALSE, warning=FALSE, paged.print=FALSE}
benchmark_dir = "./"

library(plyr)
library(ggpubr)
source(paste0(benchmark_dir,"bin/corr-analysis.R"))
```

## Data
We used the gene expression matrix to count the number of expressed genes per cell. Expressed gene -> one or more IMUs detected from this gene.

```{r data}
files_list <- list.files("output/data", recursive = T, pattern = ".Rdata", full.names=TRUE)

for (dt in c("QC.per.cell", "total_UMI")){
    li.tmp <- list()
    idx <- grep(dt, files_list)
          for (file in files_list[idx]){
              data_type <- strsplit(file, split ="/")[[1]][3]
              li.tmp[[data_type]] <- as.data.frame(readRDS(file))
              }
    assign(dt, li.tmp)
}

data.QC <- ldply(QC.per.cell, rbind)
```

```{r}
box.cor <-  data.QC %>%
  mutate(workflow = factor(workflow, levels=c( 'scNanoGPS', 'Sockeye', 'Sicelore2.1', 'FLAMES', 'Snuupy', 'Sicelore'))) %>%
  ggplot(aes(fill=data, y=correlation, x=workflow))+
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    theme_classic()+
    scale_fill_manual(values = c("#B276B2", "#FAA43A", "#5DA5DA")) + # Change the colors 5DA5DA
    labs(x = "", y = "Pearson's correlation", title ="Illumina gene expression correlation" ) +
    theme(axis.text.x =  element_text(colour = "black"),
          axis.text = element_text(face="bold"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank()) 
```


### Number of detected UMI per cell (MinION run) 

**Total Number of UMI per cell for common cells for different preprocessing workflows against Cell Ranger on the PromethION dataset. The Pearson correlation coefficient is ploted**

```{r}
count.per.cell <- rbind(QC.per.cell$MinION, QC.per.cell$PromethION, QC.per.cell$Spatial)
```


```{r}
#colnames(count.per.cell) <- c("Illumina.upc", "variable", "UPC", "Illumina.gpc", "GPC", "data")
box.umi <- count.per.cell %>%
  mutate(workflow = factor(workflow, levels=c( 'scNanoGPS', 'Sockeye', 'Sicelore2.1', 'FLAMES', 'Snuupy', 'Sicelore'))) %>%
  ggplot(aes(fill=data, y=umi, x=workflow))+
    ylim(c(0,25000))+
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    theme_classic()+
    scale_fill_manual(values = c("#B276B2", "#FAA43A", "#5DA5DA")) + # Change the colors 5DA5DA
    #scale_fill_brewer(palette = "Dark2")+
    labs(x = "", y = "UMI count", title ="No. of UMI per cell") +
    theme(axis.text.x =  element_text(colour = "black"),
          axis.text = element_text(face="bold"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank()) 
```

```{r}
box.genes <- count.per.cell %>%
  #mutate(day = fct_reorder(variable, GPC)) %>%
  mutate(workflow = factor(workflow, levels=c( 'scNanoGPS', 'Sockeye', 'Sicelore2.1', 'FLAMES', 'Snuupy', 'Sicelore'))) %>%
  ggplot(aes(fill=data, y=gene, x=workflow))+
    ylim(c(0,5000))+
    geom_boxplot(width = 0.5, outlier.shape = NA) +
    theme_classic()+
    scale_fill_manual(values = c("#B276B2", "#FAA43A", "#5DA5DA")) + # Change the colors 5DA5DA
    #scale_fill_brewer(palette = "Dark2")+
    labs(x = "", y = "Gene count", title ="No. of genes per cell") +
    theme(axis.text.x =  element_text(colour = "black"),
          axis.text = element_text(face="bold"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank())
```

```{r}
df_UMItotal <- rbind(total_UMI$MinION, total_UMI$PromethION, total_UMI$Spatial)
bar.total <-  df_UMItotal %>%
  #mutate(day = fct_reorder(pipe, total_UMI)) %>%
  mutate(pipe = factor(pipe, levels=c( 'scNanoGPS', 'Sockeye', 'Sicelore2.1', 'FLAMES', 'Snuupy', 'Sicelore'))) %>%
  ggplot(aes(fill=data, y=total_UMI, x=pipe))+
    geom_bar(stat="identity", position=position_dodge()) +
    theme_classic()+
    scale_fill_manual(values = c("#B276B2", "#FAA43A", "#5DA5DA")) + # Change the colors 5DA5DA
    #scale_fill_brewer(palette = "Dark2")+
    labs(x = "", y = "", title ="No. total of UMI") +
    scale_y_continuous(labels = scales::number_format(scale = 1e-6, suffix = "M")) +#, limits = c(0,5000000))+
    theme(axis.text.x =  element_text(colour = "black"),
          axis.text = element_text(face="bold"),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank())
```

```{r fig.height=4, fig.width=15}
ggarrange(box.umi+coord_flip(), box.genes+coord_flip(), box.cor+coord_flip(), bar.total+coord_flip(), ncol = 4, common.legend =T, legend="bottom")
```

```{r}
count.per.cell$log.umi <- log(count.per.cell$umi)
count.per.cell$log.gene <- log(count.per.cell$gene)

count.per.cell$log.umi.illumina <- log(count.per.cell$umi.illumina )
count.per.cell$log.gene.illumina <- log(count.per.cell$gene.illumina)
```

```{r fig.height=6, fig.width=14}
custom_order <-  c('Sicelore', 'Snuupy', 'FLAMES', 'Sicelore2.1', 'Sockeye', 'scNanoGPS')

# Réorganisez les niveaux de la variable "variable" en utilisant l'ordre personnalisé
count.per.cell$workflow <- factor(count.per.cell$workflow, levels = custom_order)

scater.upc <- ggscatter(count.per.cell, x = "log.umi.illumina", y = "log.umi", size = 0.2,
                                                    color = "data") +#,
                                                    # palette = "Set2",
                                                    # add = "reg.line",
                                                    #ylim = c(0, 100000),
                                                    #xlim = c(0, 100000)) +
          facet_wrap(~workflow) + 
          geom_abline(intercept = 0, slope = 1, color = "gray") + # Add confidence interval
          stat_cor(aes(color = data, label = after_stat(r.label)), label.y = c(8.7,10,11.3)) + #c(90000, 99000)) +
          #stat_cor(aes(color = data, label =  after_stat(r.label)),
          #                          label.y = c(81000, 90000, 99000))+ 
          facet_grid("UMI per cell" ~ workflow, scales="free_y") +  # scales = "free_x", space = "free_x", switch = "x")+
          #scale_y_continuous(labels = scales::number_format(scale = 1e-3, suffix = "k")) + # Change y-axis labels to "k"
          #scale_x_continuous(labels = scales::number_format(scale = 1e-3, suffix = "k")) + # Change x-axis labels to "k"
          xlab("") +  # Change x-axis label
          ylab("Nanopore") +  # Change y-axis label
          scale_color_manual(values = c("#B276B2", "#FAA43A", "#5DA5DA")) +  # Change the colors to 5DA5DA
          scale_fill_brewer(palette = "Dark2")+
          theme(axis.text.x = element_text(size = 8),  # Augmenter la taille des étiquettes x
                axis.text.y = element_text(size = 8),  # Augmenter la taille des étiquettes y
                strip.background = element_rect(fill = "transparent"),
                strip.text.x = element_text(size = 8)
        ) 

scater.gpc <- ggscatter(count.per.cell, x = "log.gene.illumina", y = "log.gene", size = 0.2,
                                                    color = "data") +#,
                                                    # palette = "Set2",
                                                    # add = "reg.line",
                                                    #ylim = c(0, 15000),
                                                    #xlim = c(0, 15000), ) +
          facet_wrap(~workflow) +
          geom_abline(intercept = 0, slope = 1, color = "gray") + # Add confidence interval
          stat_cor(aes(color = data, label = after_stat(r.label)), label.y = c(8,9,10)) + #c(90000, 99000)) +
          #stat_cor(aes(color = data, label =  after_stat(r.label)),
                                  #  label.y = c(12500, 13700, 14900))+ 
          facet_grid( "UMI per cell" ~ workflow, scales="free_y") +  # scales = "free_x", space = "free_x", switch = "x")+
          #scale_y_continuous(labels = scales::number_format(scale = 1e-3, suffix = "k")) + # Change y-axis labels to "k"
          #scale_x_continuous(labels = scales::number_format(scale = 1e-3, suffix = "k")) + # Change x-axis labels to "k"
          xlab("Illumina") +  # Change x-axis label
          ylab("Nanopore") +  # Change y-axis label
          scale_color_manual(values = c("#B276B2", "#FAA43A", "#5DA5DA")) +  # Change the colors to 5DA5DA
          #scale_fill_brewer(palette = "Dark2")+
          theme(axis.text.x = element_text(size = 8),  # Adjust x-axis text size
                axis.text.y = element_text(size = 8),  # Adjust y-axis text size
                strip.background = element_rect(fill = "transparent"),#,  # Make facet labels transparent
                strip.text.x = element_blank()
                #strip.text = element_text(color = "transparent")  # Make facet label text transparent
          )

ggarrange(scater.upc, scater.gpc, nrow=2, common.legend = TRUE)
```


```{r}
output_dir <- file.path("./output/pdf")

if (!dir.exists(output_dir)){
dir.create(output_dir)
} else {
    print("Dir already exists!")
}

pdf(file = paste0(output_dir, "/QC.pdf"), onefile=FALSE,
    width = 15, height = 4)
ggarrange(box.umi+coord_flip(), box.genes+coord_flip(), box.cor+coord_flip(), bar.total+coord_flip(), ncol = 4, common.legend =T, legend="bottom")
dev.off()

pdf(file = paste0(output_dir, "/scatters.pdf"), onefile=FALSE,
    width = 14, height = 6)
ggarrange(scater.upc, scater.gpc, nrow=2, common.legend = TRUE)
dev.off()
```


```{r}
sessionInfo()
```


